mlir/lib/Catalyst/Transforms/DetensorizeSCFPass.cpp

```diff
@@ -0,0 +1,163 @@
+            rewriter.replaceUsesWithIf(
+                newForOp.getRegionIterArg(index), from_elem_op->getResult(0),
+                [&](OpOperand &op) { return !isa<tensor::FromElementsOp>(op.getOwner()); });
+        }
+
+        // 5. Extract scalar tensor elements and detensorize terminator
+        auto clonedYieldOp = cast<scf::YieldOp>(newBlock.getTerminator());
+        SmallVector<Value> newYieldOperands = clonedYieldOp.getOperands();
+        rewriter.setInsertionPoint(clonedYieldOp);
+        for (std::size_t index : indices) {
+            mlir::Operation *extract_op = rewriter.create<tensor::ExtractOp>(
+                clonedYieldOp->getLoc(), clonedYieldOp->getOperand(index), ValueRange{});
+            newYieldOperands[index] = extract_op->getResult(0);
+        }
+        rewriter.create<scf::YieldOp>(newForOp.getLoc(), newYieldOperands);
+        rewriter.eraseOp(clonedYieldOp);
+
+        // 6. Retensorize results after for loop
+        rewriter.setInsertionPointAfter(forOp);
+        for (std::size_t index : indices) {
+            mlir::Value for_result = newForOp->getResult(index);
+            auto from_elem_op = rewriter.create<tensor::FromElementsOp>(
+                newForOp->getLoc(), RankedTensorType::get({}, for_result.getType()), for_result);
+            rewriter.replaceUsesWithIf(
+                forOp->getResult(index), from_elem_op->getResult(0),
+                [&](OpOperand &op) { return !isa<tensor::FromElementsOp>(op.getOwner()); });
+        }
+
+        rewriter.replaceOp(forOp, newForOp);
+    }
+
+    void detensorizeIfOp(scf::IfOp if_op, IRRewriter &rewriter)
+    {
+        auto isScalarTensor = [](Value v) {
+            if (!::llvm::isa<TensorType>(v.getType())) {
+                return false;
+            }
+            return dyn_cast<TensorType>(v.getType()).getRank() == 0;
+        };
+
+        if_op->walk([&](scf::YieldOp yield_op) {
+            // Loop over results
+            std::size_t i_result = 0;
+            for (mlir::Value operand : yield_op.getOperands()) {
+                if (isScalarTensor(operand)) {
+                    auto if_result = if_op->getResult(i_result);
+                    // Detensorize operand: extract tensor element before yielding
+                    {
+                        rewriter.setInsertionPoint(yield_op);
+                        mlir::Operation *extract_op = rewriter.create<tensor::ExtractOp>(
+                            yield_op->getLoc(), operand, ValueRange{});
+                        rewriter.replaceUsesWithIf(
+                            operand, extract_op->getResult(0),
+                            [&](OpOperand &op) { return op.getOwner() == yield_op; });
+                        const mlir::Type type = extract_op->getResult(0).getType();
+                        rewriter.modifyOpInPlace(
+                            yield_op, [&]() { yield_op->getOperand(i_result).setType(type); });
```

Comment:
> How would it be useful here? Is this to avoid changing the insertion point outside detensorizeIfOp?

So if you always set the insertion point before you create a new op then it is not needed. But imagine you outline a piece of code that creates new ops into a separate function, and that function modifies the insertion point without resetting it, then then caller may not realize the function modified the ip.
It's good practice to reset the ip to avoid this. The RAII class is used like this:
```
            PatternRewriter::InsertionGuard insertGuard(rewriter);
            rewriter.setInsertionPointAfter(...);
```
Again not mandatory if you're careful, but good practice :)